File: phoenix-scala/app/models/rules/Condition.scala
package models.rules

import com.pellucid.sealerate
import utils.ADT

case class Condition(rootObject: String,
                     field: String,
                     operator: Condition.Operator,
                     valInt: Option[Int] = None,
                     valString: Option[String] = None,
                     valBoolean: Option[Boolean] = None)

object Condition {
  sealed trait Operator

  // General operations
  case object Equals    extends Operator
  case object NotEquals extends Operator

  // Numeric operations
  case object GreaterThan         extends Operator
  case object GreaterThanOrEquals extends Operator
  case object LessThan            extends Operator
  case object LessThanOrEquals    extends Operator

  // String operations
  case object Contains    extends Operator
  case object NotContains extends Operator
  case object StartsWith  extends Operator
  case object InArray     extends Operator
  case object NotInArray  extends Operator

  object Operator extends ADT[Operator] {
    def types = sealerate.values[Operator]
  }

  // TODO (Jeff): Return an actual error, rather than just a boolean.
  def matches(comp: Int, statement: Condition): Boolean = {
    statement.valInt.fold(false) { (v: Int) ⇒
      statement.operator match {
        case Equals              ⇒ comp == v
        case NotEquals           ⇒ comp != v
        case GreaterThan         ⇒ comp > v
        case GreaterThanOrEquals ⇒ comp >= v
        case LessThan            ⇒ comp < v
        case LessThanOrEquals    ⇒ comp <= v
        case _                   ⇒ false
      }
    }
  }

  // TODO (Jeff): Make this more robust and think about things like case-sensitivity.
  def matches(comp: String, statement: Condition): Boolean =
    matches(Some(comp), statement)

  def matches(comp: Option[String], statement: Condition): Boolean = {
    comp.fold(matchAgainstEmptyStringOption(statement)) { (comp: String) ⇒
      statement.valString.fold(false) { (v: String) ⇒
        statement.operator match {
          case Equals      ⇒ comp == v
          case NotEquals   ⇒ comp != v
          case Contains    ⇒ comp.contains(v)
          case NotContains ⇒ !comp.contains(v)
          case StartsWith  ⇒ comp.startsWith(v)
          case InArray     ⇒ v.split(", ").contains(comp)
          case NotInArray  ⇒ !v.split(", ").contains(comp)
          case _           ⇒ false
        }
      }
    }
  }

  def matches(comp: Boolean, condition: Condition): Boolean = {
    condition.valBoolean.fold(false) { (v: Boolean) ⇒
      condition.operator match {
        case Equals    ⇒ comp == v
        case NotEquals ⇒ comp != v
        case _         ⇒ false
      }
    }
  }

  private def matchAgainstEmptyStringOption(statement: Condition): Boolean = {
    statement.operator match {
      case Equals      ⇒ statement.valString.isEmpty
      case NotEquals   ⇒ statement.valString.nonEmpty
      case NotContains ⇒ true
      case _           ⇒ false
    }
  }
}

File: middlewarehouse/models/rules/condition.go
package rules

import (
	"fmt"
	"strings"
)

type Condition struct {
	RootObject string      `json:"rootObject"`
	Field      string      `json:"field"`
	Operator   string      `json:"operator"`
	Value      interface{} `json:"value"`
}

func (c Condition) MatchesBool(comp bool) (bool, error) {
	var result bool
	var err error

	valBool, ok := c.Value.(bool)
	if !ok {
		return false, fmt.Errorf(errorInvalidTypeCast, "boolean")
	}

	switch c.Operator {
	case Equals:
		result = comp == valBool
	case NotEquals:
		result = comp != valBool
	default:
		err = fmt.Errorf(errorInvalidComparison, c.Operator, "boolean")
	}

	return result, err
}

func (c Condition) MatchesInt(comp int) (bool, error) {
	var result bool
	var err error

	valInt, ok := c.Value.(int)
	if !ok {
		return false, fmt.Errorf(errorInvalidTypeCast, "integer")
	}

	switch c.Operator {
	case Equals:
		result = comp == valInt
	case NotEquals:
		result = comp != valInt
	case GreaterThan:
		result = comp > valInt
	case GreaterThanOrEquals:
		result = comp >= valInt
	case LessThan:
		result = comp < valInt
	case LessThanOrEquals:
		result = comp <= valInt
	default:
		err = fmt.Errorf(errorInvalidComparison, c.Operator, "boolean")
	}

	return result, err
}

func (c Condition) MatchesString(comp string) (bool, error) {
	var result bool
	var err error

	valStr, ok := c.Value.(string)
	if !ok {
		return false, fmt.Errorf(errorInvalidTypeCast, "string")
	}

	switch c.Operator {
	case Equals:
		result = comp == valStr
	case NotEquals:
		result = comp != valStr
	case Contains:
		result = strings.Contains(comp, valStr)
	case NotContains:
		result = !strings.Contains(comp, valStr)
	case StartsWith:
		result = strings.HasPrefix(comp, valStr)
	case InArray:
		result = isInArray(comp, valStr)
	case NotInArray:
		result = !isInArray(comp, valStr)
	default:
		err = fmt.Errorf(errorInvalidComparison, c.Operator, "string")
	}

	return result, err
}

func (c Condition) MatchesUint(comp uint) (bool, error) {
	var result bool
	var err error

	valInt, ok := c.Value.(uint)
	if !ok {
		return false, fmt.Errorf(errorInvalidTypeCast, "unsigned integer")
	}
	switch c.Operator {
	case Equals:
		result = comp == valInt
	case NotEquals:
		result = comp != valInt
	case GreaterThan:
		result = comp > valInt
	case GreaterThanOrEquals:
		result = comp >= valInt
	case LessThan:
		result = comp < valInt
	case LessThanOrEquals:
		result = comp <= valInt
	default:
		err = fmt.Errorf(errorInvalidComparison, c.Operator, "boolean")
	}

	return result, err
}

func isInArray(comp string, val string) bool {
	valArray := strings.Split(val, ", ")

	for _, v := range valArray {
		if comp == v {
			return true
		}
	}

	return false
}
